### ISSUE TITLE
ContentView inherits from the wrong base Layout class

### ISSUE THREAD

**Issue by ivan-todorov-progress** [dotnet/maui#7613]
### Description

There are two parallel layout implementations in .NET MAUI:
1. The new layout implementation with the base class of `Telerik.Maui.Controls.Layout`
2. The legacy layout implementation from Xamarin.Forms with the base class of `Telerik.Maui.Controls.Compatibility.Layout`

The new layout implementation is compatible with the new MAUI handlers. The old layout implementation from Xamarin.Forms is not compatible with the handlers and should be avoided.

The problem with the `ContentView` is that it has a handler, but is still inherits from the legacy layout `Telerik.Maui.Controls.Compatibility.Layout`. This is a major source of bugs, as the usual layout logic might not be executed correctly or might not get executed at all on some platforms.

### Steps to Reproduce

Download and run the attached sample project on Windows.

[TestApp.zip](https://github.com/dotnet/maui/files/8799692/TestApp.zip)

The sample project contains a single `CustomView` inheriting from `ContentView`. The sole purpose of that `CustomView` is to override all the virtual `Measure`, `Arrange`, `Layout` etc. methods of the base class and dump some debug output when they are called. The sample applications has two additional buttons to call the `InvalidateMeasure` method on the `CustomView` of the base `VisualElement` class and via the `IView` interface.

Observe the debug output of the application. You can also put breakpoints in the overridden methods to be sure. Notice that nothing happens: nothing is called at all. All the layout logic of the `CustomView` is completely discarded.

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

N/A

**Comment by drasticactions** [comment]
Wouldn't this be an issue with the Telerik controls you're using? I'm not sure how this is a MAUI issue.

**Comment by Xyncgas** [comment]
> Wouldn't this be an issue with the Telerik controls you're using? I'm not sure how this is a MAUI issue.

:D

**Comment by ivan-todorov-progress** [comment]
> Wouldn't this be an issue with the Telerik controls you're using? I'm not sure how this is a MAUI issue.

I am one of the authors of the Telerik controls and this bug is one of the sources of such '*issues*' in our controls. The sample application I have sent uses only Microsoft controls to reproduce the bug. Please, read the post carefully and test the sample application before creating such noise in this thread.

**Comment by AswinPG** [comment]
https://github.com/dotnet/maui/issues/7668 Could this be related ?

**Comment by ivan-todorov-progress** [comment]
@AswinPG I have not debugged the MAUI source code to that extent, but the `LayoutChildren` method in `ContentPage` seems like an old paradigm from the legacy Xamarin.Forms layout implementation. The new MAUI layouts have something called `ILayoutManager`, with two methods: `Measure` and `ArrangeChildren`: https://github.com/dotnet/maui/blob/main/src/Core/src/Layouts/ILayoutManager.cs.

It is a total mess really, as the new and the legacy layout implementations define different paradigms and sometimes combining the two in the same visual tree does not work quite well, i.e.: `InvalidateMeasure` not invalidating anything, the `Measure` and `Layout` methods not always called when they should etc. Migrating everything to the new MAUI layouts would resolve many such problems and this should be top priority for the MAUI team IMHO, as a broken layout system means broken apps and many disappointed developers migrating to other UI frameworks. ;-)

**Comment by AswinPG** [comment]
> 

Yes, This really is a bummer. This is one of the core features that MAUI has to handle and it's broken, I was really excited but all these huge issues has been a disappointment. And to see that issues as important as these are not getting the importance it deserves is another bummer. 
I can confirm that these exact issues applies to controls inherited from ContentPage as well. Hope someone from the MAUI team could weigh in on this...:(

**Comment by Xyncgas** [comment]
I've been finding using webview to draw layout can already be satisfying enough, for my app to run on multiple device while having native platform capabilities.
If creating layout in MAUI is easy enough then I would use it, however you are gonna have to top all the dom events they have in the webview (onclick -- I supposed there's button in MAUI, onmouseover, onkeyup...)
but also is MAUI supporting everything well but linux

**Comment by philipag** [comment]
I agree with @ivan-todorov-progress in that this is the type of bug that is making it next to impossible to migrate complex XF apps to Maui.   This should be top priority to get fixed.

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio Enterprise 17.8.0 Preview 1.0. Can repro on windows platform with sample project.
[TestApp.zip](https://github.com/dotnet/maui/files/12594211/TestApp.zip)

**Comment by MitchBomcanhao** [comment]
not being able to use contentviews with control templates makes migrating our xamarin forms app nearly impossible.

**Comment by Alex-111** [comment]
The later this will be fixed the more apps will break. Any reasons why this is delayed?

**Comment by MitchBomcanhao** [comment]
bump!

**Comment by GalaxiaGuy** [comment]
@hartez In #19794 you say this isn't exactly a bug.

Does that mean it is intentional that `ContentView` inherits from `Compatibility.Layout`?

**Comment by hartez** [comment]
> @hartez In #19794 you say this isn't exactly a bug.
> 
> Does that mean it is intentional that `ContentView` (and `TemplatedView`) inherits from `Compatibility.Layout`?

Yes, insofar as they inherited from that class in Forms, and that base class relationship was left intact for backward compatibility reasons in the first release of .NET MAUI. 

Starting from scratch, and with no backward-compatibility goals, I doubt we'd have kept that inheritance chain. Which is why I say it isn't _exactly_ a bug; given the current capabilities and design, it certainly looks like one, but it was an intentional design choice at one point. And "fixing" it would require a breaking change.

**Comment by MitchBomcanhao** [comment]
Considering you have to remove the compatibility mode in order to get content view to even render on windows, seems like it is already broken. Might as well just fix it.

**Comment by Xyncgas** [comment]
You are losing new customer for your new framework by remaining compatibility

**Comment by jump32** [comment]
@hartez So why not have a version of ContentView with this inheritance chain under the Compatibility namespace? In the same way as you do for other controls, such as Stack, etc.

That way you can have a version in the Controls namespace that has the correct inheritance. That isn't fundamentally broken. And that, you know, actually works.

**Comment by hartez** [comment]
> @hartez So why not have a version of ContentView with this inheritance chain under the Compatibility namespace? In the same way as you do for other controls, such as Stack, etc.
> 
> That way you can have a version in the Controls namespace that has the correct inheritance. That isn't fundamentally broken. And that, you know, actually works.

I'm not clear on what you all mean by "ContentView not working", but if I take the repro project attached to this issue and update it to .NET 8, the `CustomView` class displays just fine. 

Now, if you actually put some custom measurement logic in `CustomView` and aren't just calling `base.MeasureOverride`, _at the moment_ you're required to explicitly set the value of `DesiredSize`; if you don't, the layout won't work correctly. 

So this won't work:

```csharp
protected override Size MeasureOverride(double widthConstraint, double heightConstraint)
{
	var size = new Size(200,200);
	return size;
}
```

But this will:

```csharp
protected override Size MeasureOverride(double widthConstraint, double heightConstraint)
{
	var size = new Size(200,200);
	DesiredSize = size;
	return size;
}
```

This was a design/documentation error (it wasn't documented that you needed to set `DesiredSize`, and really the root classes should just be doing that for you anyway). So after #19794, you won't have to do that anymore - the first version of `MeasureOverride` above will work. 

But that's all orthogonal to the "wrong base class" claim. The original error reported (1.5 years ago) has been fixed, and the fix didn't have much to do with the base class. Changing the base class now (even if you could without breaking any existing code)  is unlikely to address any problems you're having. If I could I would just close this issue.

**Comment by MitchBomcanhao** [comment]
@hartez what I mean by it not working is that if you have the maui compatibility mode enabled in mauiprogram.cs, then windows will not render contentview stuff. I don't know if the two problems are linked, but contentview is completely broken on windows when using compatibility mode, so this idea of not changing it because it becomes a breaking change is quite funny for everyone that is actually trying to convert existing applications and finding them to just not work at all.

**Comment by hartez** [comment]
This issue isn't about compatibility mode, though. And I'm not sure I see how changing the base class of ContentView would fix something that's broken with compatibility mode, especially if it's only broken in compatibility mode on _one platform_. That it's only broken on one platform suggests a bug in the compatibility renderer for _that platform_, not an issue with the hierarchy in the cross-platform code. 

It looks like there are other issues open for the actual problem (i.e., that ContentView doesn't work properly in compatibility mode); I would suggest the focus be on those issues (e.g., #19779), not this one. 

Unless someone has opened a PR that fixes the problem (without an API break) by changing the base class?

**Comment by PureWeen** [comment]
First step for this one
https://github.com/dotnet/maui/pull/23710

Now in .NET 10 we will probably just remove that middle layer
